### PR TITLE
2.5 - Backport date

### PR DESCRIPTION
**Pull request by markstory**
There are a few things going on here as I started small but wanted to
stop before the diff got too large.

- Added deprecations for modifing dates by their time components.
- I've disallowed diffing dissimilar types with a runtime check. The
  goal of this was to approximate the type errors that will be present
  in 3.x.
- Add deprecations for short comparison methods.

### CODE REVIEW

**Comment by othercorey** [comment]
Look ok.

File: phpstan-baseline.neon [hunk 1]
```diff
@@ -1,5 +1,10 @@
 parameters:
 	ignoreErrors:
+		-
+			message: "#^Strict comparison using \\=\\=\\= between class\\-string\\<Cake\\\\Chronos\\\\Chronos\\> and 'Cake\\\\\\\\Chronos\\\\\\\\ChronosDate' will always evaluate to false\\.$#"
+			count: 2
+			path: src/Chronos.php
```
Comment (othercorey) [inline phpstan-baseline.neon:6]: Whatever code is causing this error seems incorrect. These will never evaluate correctly.

Comment (markstory) [reply phpstan-baseline.neon:6]: These warnings were coming from the traits, and are related to the trait method having an impossible condition in some of the implementations.

Comment (othercorey) [reply phpstan-baseline.neon:6]: Ok, gotcha.

**Pull request merged**